pkg/logql/filter.go

```diff
@@ -0,0 +1,287 @@
+	}
+	return nil, false
+}
+
+// clearCapture removes capture operation as they are not used for filtering.
+func clearCapture(regs ...*syntax.Regexp) {
+	for _, r := range regs {
+		if r.Op == syntax.OpCapture {
+			*r = *r.Sub[0]
+		}
+	}
+}
+
+// simplifyAlternate simplifies, when possible, alternate regexp expressions such as:
+// (foo|bar) or (foo|(bar|buzz)).
+func simplifyAlternate(reg *syntax.Regexp) (LineFilter, bool) {
+	clearCapture(reg.Sub...)
+	// attempt to simplify the first leg
+	f, ok := simplify(reg.Sub[0])
+	if !ok {
+		return nil, false
+	}
+	// merge the rest of the legs
+	for i := 1; i < len(reg.Sub); i++ {
+		f2, ok := simplify(reg.Sub[i])
+		if !ok {
+			return nil, false
+		}
+		f = newOrFilter(f, f2)
+	}
+	return f, true
+}
+
+// simplifyConcat attempt to simplify concat operations.
+// Concat operations are either literal and star such as foo.* .*foo.* .*foo
+// which is a literalFilter.
+// Or a literal and alternates operation (see simplifyConcatAlternate), which represent a multiplication of alternates.
+// Anything else is rejected.
+func simplifyConcat(reg *syntax.Regexp, baseLiteral []byte) (LineFilter, bool) {
+	clearCapture(reg.Sub...)
+	// we support only simplication of concat operation with 3 sub expressions.
+	// for instance .*foo.*bar contains 4 subs (.*+foo+.*+bar) and can't be simplified.
+	if len(reg.Sub) > 3 {
+		return nil, false
+	}
+
+	var curr LineFilter
+	var ok bool
+	literals := 0
+	for _, sub := range reg.Sub {
+		if sub.Op == syntax.OpLiteral {
+			// only one literal is allowed.
+			if literals != 0 {
+				return nil, false
+			}
+			literals++
+			baseLiteral = append(baseLiteral, []byte(string(sub.Rune))...)
+			continue
+		}
+		// if we have an alternate we must also have a base literal to apply the concatenation with.
+		if sub.Op == syntax.OpAlternate && baseLiteral != nil {
+			if curr, ok = simplifyConcatAlternate(sub, baseLiteral, curr); !ok {
+				return nil, false
+			}
+			continue
+		}
+		if sub.Op == syntax.OpStar && sub.Sub[0].Op == syntax.OpAnyCharNotNL {
+			continue
+		}
+		return nil, false
+	}
+
+	// if we have a filter from concat alternates.
+	if curr != nil {
+		return curr, true
+	}
+
+	// if we have only a concat with literals.
+	if baseLiteral != nil {
+		return literalFilter(baseLiteral), true
+
+	}
+
+	return nil, false
+}
+
+// simplifyConcatAlternate simplifies concat alternate operations.
+// A concat alternate is found when a concat operation has a sub alternate and is preceded by a literal.
+// For instance bar|b|buzz is expressed as b(ar|(?:)|uzz) => b concat alternate(ar,(?:),uzz).
+// (?:) being an OpEmptyMatch and b being the literal to concat all alternates (ar,(?:),uzz) with.
+func simplifyConcatAlternate(reg *syntax.Regexp, literal []byte, curr LineFilter) (LineFilter, bool) {
+	for _, alt := range reg.Sub {
+		switch alt.Op {
+		case syntax.OpEmptyMatch:
+			curr = chainOrFilter(curr, literalFilter(literal))
```

Comment:
I'm not sure this is right? If `OpEmptyMatch` is supposed to match an empty string, then this will always falsely report true if we're using a bytes contains rather than a `string(literal) == ""`.

Reply:
Yeah it's hard to understand you have to look at the comments and tests but basically this regex

f|foo|foobar translate into f ((?:)|oo|oobar) once parsed. which is concat(f,alternate(empty,oo,oobar) empty means we take the base literal without concatenating it to anything so we get f | foo | foobar.

Reply:
so the baseliteral in this case is `f`

Reply:
Ah, I understand now. Thanks!